[BUG] value too long for type character varying(1024)

### Description

I have some (mostly multipage) documents that show up in the not processed queue. When I redo the scan of a failing document ist also throws the same error:

value too long for type character varying(1024)

But some of my scanned documents mostly small one page documents are processed as expected.

Any ideas what is worng or what I did wrong ?

### Steps to reproduce

1. Scan a document with more than one page
2. expect to see it as document but it is on the error list:

scan_017562.pdf: The following error occurred while storing document scan_017562.pdf after parsing: value too long for type character varying(1024)

and still persent in the consume folder.

### Webserver logs

```bash
[2026-03-03 11:52:53,107] [DEBUG] [paperless.tasks] Executing plugin ConsumerPreflightPlugin

[2026-03-03 11:52:53,125] [INFO] [paperless.tasks] ConsumerPreflightPlugin completed with no message

[2026-03-03 11:52:53,125] [DEBUG] [paperless.tasks] Skipping plugin CollatePlugin

[2026-03-03 11:52:53,128] [DEBUG] [paperless.tasks] Executing plugin BarcodePlugin

[2026-03-03 11:52:53,128] [DEBUG] [paperless.barcodes] Scanning for barcodes using PYZBAR

[2026-03-03 11:52:53,129] [DEBUG] [paperless.barcodes] PDF has 2 pages

[2026-03-03 11:52:53,129] [DEBUG] [paperless.barcodes] Processing page 0

[2026-03-03 11:52:53,599] [DEBUG] [paperless.barcodes] Image is at /tmp/paperless/tmpp87q9ppv/barcodee74co91d/50fe6be2-6043-48a6-981f-5a42ffeb9c66-1.ppm

[2026-03-03 11:52:54,017] [DEBUG] [paperless.barcodes] Processing page 1

[2026-03-03 11:52:54,464] [DEBUG] [paperless.barcodes] Image is at /tmp/paperless/tmpp87q9ppv/barcodee74co91d/89b677b6-5802-4a81-a4a9-367cbc184266-2.ppm

[2026-03-03 11:52:54,865] [INFO] [paperless.tasks] BarcodePlugin completed with no message

[2026-03-03 11:52:54,865] [DEBUG] [paperless.tasks] Executing plugin WorkflowTriggerPlugin

[2026-03-03 11:52:54,874] [INFO] [paperless.tasks] WorkflowTriggerPlugin completed with:

[2026-03-03 11:52:54,874] [DEBUG] [paperless.tasks] Executing plugin ConsumeTaskPlugin

[2026-03-03 11:52:54,875] [INFO] [paperless.consumer] Consuming scan_017562.pdf

[2026-03-03 11:52:54,877] [DEBUG] [paperless.consumer] Detected mime type: application/pdf

[2026-03-03 11:52:54,877] [INFO] [paperless.consumer] Executing pre-consume script /usr/src/paperless/scripts/pre-consume.sh

[2026-03-03 11:52:54,889] [INFO] [paperless.consumer] /usr/src/paperless/scripts/pre-consume.sh exited 0

[2026-03-03 11:52:54,889] [INFO] [paperless.consumer] /usr/src/paperless/scripts/pre-consume.sh stdout:

[2026-03-03 11:52:54,889] [INFO] [paperless.consumer] A document will be consumed.  I know the

[2026-03-03 11:52:54,890] [INFO] [paperless.consumer] following additional information about it:

[2026-03-03 11:52:54,890] [INFO] [paperless.consumer]

[2026-03-03 11:52:54,890] [INFO] [paperless.consumer] * Archive Path:

[2026-03-03 11:52:54,890] [INFO] [paperless.consumer] * Source Path: /usr/src/paperless/consume/scan_017562.pdf

[2026-03-03 11:52:54,890] [INFO] [paperless.consumer] * dirname: /usr/src/paperless/consume

[2026-03-03 11:52:54,890] [INFO] [paperless.consumer] ' slower : scan_017562.pdf

[2026-03-03 11:52:54,890] [INFO] [paperless.consumer] * snormal: scan_017562.pdf

[2026-03-03 11:52:54,898] [DEBUG] [paperless.consumer] Parser: RasterisedDocumentParser

[2026-03-03 11:52:54,902] [DEBUG] [paperless.consumer] Parsing scan_017562.pdf...

[2026-03-03 11:52:54,926] [INFO] [paperless.parsing.tesseract] pdftotext exited 0

[2026-03-03 11:52:55,268] [DEBUG] [paperless.parsing.tesseract] Calling OCRmyPDF with args: {'input_file': PosixPath('/tmp/paperless/paperless-ngxpnkik_e5/scan_017562.pdf'), 'output_file': PosixPath('/tmp/paperless/paperless-n5rqidfl/archive.pdf'), 'use_threads': True, 'jobs': 56, 'language': 'deu+eng', 'output_type': 'pdfa', 'progress_bar': False, 'color_conversion_strategy': 'RGB', 'skip_text': True, 'clean': True, 'deskew': True, 'rotate_pages': True, 'rotate_pages_threshold': 12.0, 'sidecar': PosixPath('/tmp/paperless/paperless-n5rqidfl/sidecar.txt'), 'invalidate_digital_signatures': True}

[2026-03-03 11:52:55,483] [INFO] [ocrmypdf._pipelines.ocr] Start processing 2 pages concurrently

[2026-03-03 11:52:57,450] [INFO] [ocrmypdf._pipeline] page is facing ⇧, confidence 8.03 - no change

[2026-03-03 11:52:57,498] [INFO] [ocrmypdf._pipeline] page is facing ⇧, confidence 9.53 - no change

[2026-03-03 11:53:09,251] [WARNING] [ocrmypdf._exec.tesseract] [tesseract] lots of diacritics - possibly poor OCR

[2026-03-03 11:53:10,127] [INFO] [ocrmypdf._pipelines.ocr] Postprocessing...

[2026-03-03 11:53:11,902] [INFO] [ocrmypdf._pipeline] Image optimization ratio: 1.28 savings: 22.0%

[2026-03-03 11:53:11,903] [INFO] [ocrmypdf._pipeline] Total file size ratio: 0.89 savings: -12.7%

[2026-03-03 11:53:11,905] [INFO] [ocrmypdf._pipelines._common] Output file is a PDF/A-2b (as expected)

[2026-03-03 11:53:12,067] [DEBUG] [paperless.parsing.tesseract] Using text from sidecar file

[2026-03-03 11:53:12,067] [DEBUG] [paperless.consumer] Generating thumbnail for scan_017562.pdf...

[2026-03-03 11:53:12,075] [DEBUG] [paperless.parsing] Execute: convert -density 300 -scale 500x5000> -alpha remove -strip -auto-orient -define pdf:use-cropbox=true /tmp/paperless/paperless-n5rqidfl/archive.pdf[0] /tmp/paperless/paperless-n5rqidfl/convert.webp

[2026-03-03 11:53:13,668] [INFO] [paperless.parsing] convert exited 0

[2026-03-03 11:53:16,714] [DEBUG] [paperless.consumer] Saving record to database

[2026-03-03 11:53:16,714] [DEBUG] [paperless.consumer] Creation date from parse_date: 2025-07-02 00:00:00+02:00

[2026-03-03 11:53:17,820] [INFO] [paperless.handlers] Assigning correspondent mi:kochkach to 2025-07-02T00:00:00+02:00 scan_017562

[2026-03-03 11:53:17,828] [INFO] [paperless.handlers] Assigning document type Uebergabeprotokoll to 2025-07-02T00:00:00+02:00 mi:kochkach scan_017562

[2026-03-03 11:53:17,841] [INFO] [paperless.handlers] Assigning storage path PARTNER to 2025-07-02T00:00:00+02:00 mi:kochkach scan_017562

[2026-03-03 11:53:18,017] [DEBUG] [paperless.index] Index updated for document 9707.

[2026-03-03 11:53:18,122] [INFO] [paperless.matching] Document did not match Workflow: Zaehlerstand

[2026-03-03 11:53:18,122] [DEBUG] [paperless.matching] Document doc type Uebergabeprotokoll does not match Zaehlerstand

[2026-03-03 11:53:18,139] [INFO] [paperless.matching] Document did not match Workflow: SEPA-Lastschrift

[2026-03-03 11:53:18,140] [DEBUG] [paperless.matching] Document doc type Uebergabeprotokoll does not match SEPA-Lastschriftmandat

[2026-03-03 11:53:18,157] [INFO] [paperless.matching] Document did not match Workflow: Rechnung

[2026-03-03 11:53:18,157] [DEBUG] [paperless.matching] Document doc type Uebergabeprotokoll does not match Rechnung

[2026-03-03 11:53:18,241] [ERROR] [paperless.consumer] The following error occurred while storing document scan_017562.pdf after parsing: value too long for type character varying(1024)

Traceback (most recent call last):

  File "/usr/local/lib/python3.12/site-packages/django/db/backends/utils.py", line 105, in _execute

    return self.cursor.execute(sql, params)

           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/usr/local/lib/python3.12/site-packages/psycopg/cursor.py", line 97, in execute

    raise ex.with_traceback(None)

psycopg.errors.StringDataRightTruncation: value too long for type character varying(1024)


The above exception was the direct cause of the following exception:


Traceback (most recent call last):

  File "/usr/local/lib/python3.12/site-packages/asgiref/sync.py", line 298, in main_wrap

    raise exc_info[1]

  File "/usr/src/paperless/src/documents/consumer.py", line 533, in run

    document.save()

  File "/usr/local/lib/python3.12/site-packages/django/db/models/base.py", line 902, in save

    self.save_base(

  File "/usr/local/lib/python3.12/site-packages/django/db/models/base.py", line 1008, in save_base

    updated = self._save_table(

              ^^^^^^^^^^^^^^^^^

  File "/usr/local/lib/python3.12/site-packages/django/db/models/base.py", line 1138, in _save_table

    updated = self._do_update(

              ^^^^^^^^^^^^^^^^

  File "/usr/local/lib/python3.12/site-packages/django/db/models/base.py", line 1203, in _do_update

    return filtered._update(values) > 0

           ^^^^^^^^^^^^^^^^^^^^^^^^

  File "/usr/local/lib/python3.12/site-packages/django/db/models/query.py", line 1286, in _update

    return query.get_compiler(self.db).execute_sql(ROW_COUNT)

           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/usr/local/lib/python3.12/site-packages/django/db/models/sql/compiler.py", line 2060, in execute_sql

    row_count = super().execute_sql(result_type)

                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/usr/local/lib/python3.12/site-packages/django/db/models/sql/compiler.py", line 1623, in execute_sql

    cursor.execute(sql, params)

  File "/usr/local/lib/python3.12/site-packages/django/db/backends/utils.py", line 79, in execute

    return self._execute_with_wrappers(

           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/usr/local/lib/python3.12/site-packages/django/db/backends/utils.py", line 92, in _execute_with_wrappers

    return executor(sql, params, many, context)

           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/usr/local/lib/python3.12/site-packages/django/db/backends/utils.py", line 100, in _execute

    with self.db.wrap_database_errors:

         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/usr/local/lib/python3.12/site-packages/django/db/utils.py", line 91, in __exit__

    raise dj_exc_value.with_traceback(traceback) from exc_value

  File "/usr/local/lib/python3.12/site-packages/django/db/backends/utils.py", line 105, in _execute

    return self.cursor.execute(sql, params)

           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/usr/local/lib/python3.12/site-packages/psycopg/cursor.py", line 97, in execute

    raise ex.with_traceback(None)

django.db.utils.DataError: value too long for type character varying(1024)

[2026-03-03 11:53:18,246] [DEBUG] [paperless.parsing.tesseract] Deleting directory /tmp/paperless/paperless-n5rqidfl

[2026-03-03 11:53:18,247] [ERROR] [paperless.tasks] ConsumeTaskPlugin failed: scan_017562.pdf: The following error occurred while storing document scan_017562.pdf after parsing: value too long for type character varying(1024)

Traceback (most recent call last):

  File "/usr/local/lib/python3.12/site-packages/django/db/backends/utils.py", line 105, in _execute

    return self.cursor.execute(sql, params)

           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/usr/local/lib/python3.12/site-packages/psycopg/cursor.py", line 97, in execute

    raise ex.with_traceback(None)

psycopg.errors.StringDataRightTruncation: value too long for type character varying(1024)


The above exception was the direct cause of the following exception:


Traceback (most recent call last):

  File "/usr/local/lib/python3.12/site-packages/asgiref/sync.py", line 298, in main_wrap

    raise exc_info[1]

  File "/usr/src/paperless/src/documents/consumer.py", line 533, in run

    document.save()

  File "/usr/local/lib/python3.12/site-packages/django/db/models/base.py", line 902, in save

    self.save_base(

  File "/usr/local/lib/python3.12/site-packages/django/db/models/base.py", line 1008, in save_base

    updated = self._save_table(

              ^^^^^^^^^^^^^^^^^

  File "/usr/local/lib/python3.12/site-packages/django/db/models/base.py", line 1138, in _save_table

    updated = self._do_update(

              ^^^^^^^^^^^^^^^^

  File "/usr/local/lib/python3.12/site-packages/django/db/models/base.py", line 1203, in _do_update

    return filtered._update(values) > 0

           ^^^^^^^^^^^^^^^^^^^^^^^^

  File "/usr/local/lib/python3.12/site-packages/django/db/models/query.py", line 1286, in _update

    return query.get_compiler(self.db).execute_sql(ROW_COUNT)

           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/usr/local/lib/python3.12/site-packages/django/db/models/sql/compiler.py", line 2060, in execute_sql

    row_count = super().execute_sql(result_type)

                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/usr/local/lib/python3.12/site-packages/django/db/models/sql/compiler.py", line 1623, in execute_sql

    cursor.execute(sql, params)

  File "/usr/local/lib/python3.12/site-packages/django/db/backends/utils.py", line 79, in execute

    return self._execute_with_wrappers(

           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/usr/local/lib/python3.12/site-packages/django/db/backends/utils.py", line 92, in _execute_with_wrappers

    return executor(sql, params, many, context)

           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/usr/local/lib/python3.12/site-packages/django/db/backends/utils.py", line 100, in _execute

    with self.db.wrap_database_errors:

         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/usr/local/lib/python3.12/site-packages/django/db/utils.py", line 91, in __exit__

    raise dj_exc_value.with_traceback(traceback) from exc_value

  File "/usr/local/lib/python3.12/site-packages/django/db/backends/utils.py", line 105, in _execute

    return self.cursor.execute(sql, params)

           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/usr/local/lib/python3.12/site-packages/psycopg/cursor.py", line 97, in execute

    raise ex.with_traceback(None)

django.db.utils.DataError: value too long for type character varying(1024)


The above exception was the direct cause of the following exception:


Traceback (most recent call last):

  File "/usr/src/paperless/src/documents/tasks.py", line 183, in consume_file

    msg = plugin.run()

          ^^^^^^^^^^^^

  File "/usr/src/paperless/src/documents/consumer.py", line 557, in run

    self._fail(

  File "/usr/src/paperless/src/documents/consumer.py", line 148, in _fail

    raise ConsumerError(f"{self.filename}: {log_message or message}") from exception

documents.consumer.ConsumerError: scan_017562.pdf: The following error occurred while storing document scan_017562.pdf after parsing: value too long for type character varying(1024)
```

### Browser logs

```bash

```

### Services logs

```bash

```

### Paperless-ngx version

2.20.9

### Host OS

debian docker

### Installation method

Docker - official image

### System status

```json
{
    "pngx_version": "2.20.9",
    "server_os": "Linux-6.8.12-10-pve-x86_64-with-glibc2.41",
    "install_type": "docker",
    "storage": {
        "total": 52521566208,
        "available": 18455863296
    },
    "database": {
        "type": "postgresql",
        "url": "paperless",
        "status": "OK",
        "error": null,
        "migration_status": {
            "latest_migration": "documents.1075_workflowaction_order",
            "unapplied_migrations": []
        }
    },
    "tasks": {
        "redis_url": "redis://broker:6379",
        "redis_status": "OK",
        "redis_error": null,
        "celery_status": "OK",
        "celery_url": "celery@f752ae293fec",
        "celery_error": null,
        "index_status": "OK",
        "index_last_modified": "2026-03-03T11:53:18.038772+01:00",
        "index_error": null,
        "classifier_status": "OK",
        "classifier_last_trained": "2026-03-03T10:05:16.160497Z",
        "classifier_error": null,
        "sanity_check_status": "ERROR",
        "sanity_check_last_run": "2026-02-28T23:30:42.718225Z",
        "sanity_check_error": "778 issues found. Check logs for details."
    },
    "websocket_connected": "ERROR"
}
```

### Browser

Chrome

### Configuration changes

_No response_

### Please confirm the following

- [x] I believe this issue is a bug that affects all users of Paperless-ngx, not something specific to my installation.
- [x] This issue is not about the OCR or archive creation of a specific file(s). Otherwise, please see above regarding OCR tools.
- [x] I have already searched for relevant existing issues and discussions before opening this report.
- [x] I have updated the title field above with a concise description.

--- a/src/documents/consumer.py
+++ b/src/documents/consumer.py
@@ -50,8 +50,6 @@ from documents.utils import copy_file_with_basic_stats
 from documents.utils import run_subprocess
 from paperless_mail.parsers import MailDocumentParser
 
-MAX_STORED_FILENAME_LENGTH = 1024
-
 
 class WorkflowTriggerPlugin(
     NoCleanupPluginMixin,
@@ -497,7 +495,10 @@ class ConsumerPlugin(
                 # place. If this fails, we'll also rollback the transaction.
                 with FileLock(settings.MEDIA_LOCK):
                     generated_filename = generate_unique_filename(document)
-                    if len(str(generated_filename)) > MAX_STORED_FILENAME_LENGTH:
+                    if (
+                        len(str(generated_filename))
+                        > Document.MAX_STORED_FILENAME_LENGTH
+                    ):
                         self.log.warning(
                             "Generated source filename exceeds db path limit, falling back to default naming",
                         )
@@ -529,7 +530,7 @@ class ConsumerPlugin(
                         )
                         if (
                             len(str(generated_archive_filename))
-                            > MAX_STORED_FILENAME_LENGTH
+                            > Document.MAX_STORED_FILENAME_LENGTH
                         ):
                             self.log.warning(
                                 "Generated archive filename exceeds db path limit, falling back to default naming",

--- a/src/documents/consumer.py
+++ b/src/documents/consumer.py
@@ -19,6 +19,7 @@ from documents.classifier import load_classifier
 from documents.data_models import ConsumableDocument
 from documents.data_models import DocumentMetadataOverrides
 from documents.file_handling import create_source_path_directory
+from documents.file_handling import generate_filename
 from documents.file_handling import generate_unique_filename
 from documents.loggers import LoggingMixin
 from documents.models import Correspondent
@@ -48,6 +49,8 @@ from documents.utils import copy_basic_file_stats
 from documents.utils import copy_file_with_basic_stats
 from documents.utils import run_subprocess
 from paperless_mail.parsers import MailDocumentParser
+
+MAX_STORED_FILENAME_LENGTH = 1024
 
 
 class WorkflowTriggerPlugin(
@@ -493,7 +496,16 @@ class ConsumerPlugin(
                 # After everything is in the database, copy the files into
                 # place. If this fails, we'll also rollback the transaction.
                 with FileLock(settings.MEDIA_LOCK):
-                    document.filename = generate_unique_filename(document)
+                    generated_filename = generate_unique_filename(document)
+                    if len(str(generated_filename)) > MAX_STORED_FILENAME_LENGTH:
+                        self.log.warning(
+                            "Generated source filename exceeds db path limit, falling back to default naming",
+                        )
+                        generated_filename = generate_filename(
+                            document,
+                            use_format=False,
+                        )
+                    document.filename = generated_filename
                     create_source_path_directory(document.source_path)
 
                     self._write(
@@ -511,10 +523,23 @@ class ConsumerPlugin(
                     )
 
                     if archive_path and Path(archive_path).is_file():
-                        document.archive_filename = generate_unique_filename(
+                        generated_archive_filename = generate_unique_filename(
                             document,
                             archive_filename=True,
                         )
+                        if (
+                            len(str(generated_archive_filename))
+                            > MAX_STORED_FILENAME_LENGTH
+                        ):
+                            self.log.warning(
+                                "Generated archive filename exceeds db path limit, falling back to default naming",
+                            )
+                            generated_archive_filename = generate_filename(
+                                document,
+                                archive_filename=True,
+                                use_format=False,
+                            )
+                        document.archive_filename = generated_archive_filename
                         create_source_path_directory(document.archive_path)
                         self._write(
                             document.storage_type,

--- a/src/documents/file_handling.py
+++ b/src/documents/file_handling.py
@@ -128,17 +128,21 @@ def generate_filename(
     counter=0,
     append_gpg=True,
     archive_filename=False,
+    use_format=True,
 ) -> Path:
     base_path: Path | None = None
 
     # Determine the source of the format string
-    if doc.storage_path is not None:
-        filename_format = doc.storage_path.path
-    elif settings.FILENAME_FORMAT is not None:
-        # Maybe convert old to new style
-        filename_format = convert_format_str_to_template_format(
-            settings.FILENAME_FORMAT,
-        )
+    if use_format:
+        if doc.storage_path is not None:
+            filename_format = doc.storage_path.path
+        elif settings.FILENAME_FORMAT is not None:
+            # Maybe convert old to new style
+            filename_format = convert_format_str_to_template_format(
+                settings.FILENAME_FORMAT,
+            )
+        else:
+            filename_format = None
     else:
         filename_format = None
 

--- a/src/documents/models.py
+++ b/src/documents/models.py
@@ -160,6 +160,7 @@ class Document(SoftDeleteModel, ModelWithOwner):
         (STORAGE_TYPE_UNENCRYPTED, _("Unencrypted")),
         (STORAGE_TYPE_GPG, _("Encrypted with GNU Privacy Guard")),
     )
+    MAX_STORED_FILENAME_LENGTH: Final[int] = 1024
 
     correspondent = models.ForeignKey(
         Correspondent,
@@ -267,7 +268,7 @@ class Document(SoftDeleteModel, ModelWithOwner):
 
     filename = models.FilePathField(
         _("filename"),
-        max_length=1024,
+        max_length=MAX_STORED_FILENAME_LENGTH,
         editable=False,
         default=None,
         unique=True,
@@ -277,7 +278,7 @@ class Document(SoftDeleteModel, ModelWithOwner):
 
     archive_filename = models.FilePathField(
         _("archive filename"),
-        max_length=1024,
+        max_length=MAX_STORED_FILENAME_LENGTH,
         editable=False,
         default=None,
         unique=True,
@@ -287,7 +288,7 @@ class Document(SoftDeleteModel, ModelWithOwner):
 
     original_filename = models.CharField(
         _("original filename"),
-        max_length=1024,
+        max_length=MAX_STORED_FILENAME_LENGTH,
         editable=False,
         default=None,
         unique=False,

--- a/src/documents/signals/handlers.py
+++ b/src/documents/signals/handlers.py
@@ -59,7 +59,6 @@ if TYPE_CHECKING:
     from documents.data_models import DocumentMetadataOverrides
 
 logger = logging.getLogger("paperless.handlers")
-MAX_STORED_FILENAME_LENGTH = 1024
 
 
 def add_inbox_tags(sender, document: Document, logging_group=None, **kwargs):
@@ -464,11 +463,11 @@ def update_filename_and_move_files(
             move_original = False
 
             candidate_filename = generate_filename(instance)
-            if len(str(candidate_filename)) > MAX_STORED_FILENAME_LENGTH:
+            if len(str(candidate_filename)) > Document.MAX_STORED_FILENAME_LENGTH:
                 msg = (
                     f"Document {instance!s}: Generated filename exceeds db path "
                     f"limit ({len(str(candidate_filename))} > "
-                    f"{MAX_STORED_FILENAME_LENGTH}): {candidate_filename!s}"
+                    f"{Document.MAX_STORED_FILENAME_LENGTH}): {candidate_filename!s}"
                 )
                 logger.warning(msg)
                 raise CannotMoveFilesException(msg)
@@ -497,11 +496,11 @@ def update_filename_and_move_files(
 
             if instance.has_archive_version:
                 archive_candidate = generate_filename(instance, archive_filename=True)
-                if len(str(archive_candidate)) > MAX_STORED_FILENAME_LENGTH:
+                if len(str(archive_candidate)) > Document.MAX_STORED_FILENAME_LENGTH:
                     msg = (
                         f"Document {instance!s}: Generated archive filename exceeds "
                         f"db path limit ({len(str(archive_candidate))} > "
-                        f"{MAX_STORED_FILENAME_LENGTH}): {archive_candidate!s}"
+                        f"{Document.MAX_STORED_FILENAME_LENGTH}): {archive_candidate!s}"
                     )
                     logger.warning(msg)
                     raise CannotMoveFilesException(msg)

--- a/src/documents/signals/handlers.py
+++ b/src/documents/signals/handlers.py
@@ -59,6 +59,7 @@ if TYPE_CHECKING:
     from documents.data_models import DocumentMetadataOverrides
 
 logger = logging.getLogger("paperless.handlers")
+MAX_STORED_FILENAME_LENGTH = 1024
 
 
 def add_inbox_tags(sender, document: Document, logging_group=None, **kwargs):
@@ -460,8 +461,18 @@ def update_filename_and_move_files(
 
             old_filename = instance.filename
             old_source_path = instance.source_path
+            move_original = False
 
             candidate_filename = generate_filename(instance)
+            if len(str(candidate_filename)) > MAX_STORED_FILENAME_LENGTH:
+                msg = (
+                    f"Document {instance!s}: Generated filename exceeds db path "
+                    f"limit ({len(str(candidate_filename))} > "
+                    f"{MAX_STORED_FILENAME_LENGTH}): {candidate_filename!s}"
+                )
+                logger.warning(msg)
+                raise CannotMoveFilesException(msg)
+
             candidate_source_path = (
                 settings.ORIGINALS_DIR / candidate_filename
             ).resolve()
@@ -482,9 +493,18 @@ def update_filename_and_move_files(
 
             old_archive_filename = instance.archive_filename
             old_archive_path = instance.archive_path
+            move_archive = False
 
             if instance.has_archive_version:
                 archive_candidate = generate_filename(instance, archive_filename=True)
+                if len(str(archive_candidate)) > MAX_STORED_FILENAME_LENGTH:
+                    msg = (
+                        f"Document {instance!s}: Generated archive filename exceeds "
+                        f"db path limit ({len(str(archive_candidate))} > "
+                        f"{MAX_STORED_FILENAME_LENGTH}): {archive_candidate!s}"
+                    )
+                    logger.warning(msg)
+                    raise CannotMoveFilesException(msg)
                 archive_candidate_path = (
                     settings.ARCHIVE_DIR / archive_candidate
                 ).resolve()

--- a/src/documents/signals/handlers.py
+++ b/src/documents/signals/handlers.py
@@ -462,6 +462,10 @@ def update_filename_and_move_files(
             old_source_path = instance.source_path
             move_original = False
 
+            old_archive_filename = instance.archive_filename
+            old_archive_path = instance.archive_path
+            move_archive = False
+
             candidate_filename = generate_filename(instance)
             if len(str(candidate_filename)) > Document.MAX_STORED_FILENAME_LENGTH:
                 msg = (
@@ -489,10 +493,6 @@ def update_filename_and_move_files(
             # Need to convert to string to be able to save it to the db
             instance.filename = str(new_filename)
             move_original = old_filename != instance.filename
-
-            old_archive_filename = instance.archive_filename
-            old_archive_path = instance.archive_path
-            move_archive = False
 
             if instance.has_archive_version:
                 archive_candidate = generate_filename(instance, archive_filename=True)

--- a/src/documents/tests/test_consumer.py
+++ b/src/documents/tests/test_consumer.py
@@ -633,8 +633,12 @@ class TestConsumer(
 
         self._assert_first_last_send_progress()
 
-    @override_settings(FILENAME_FORMAT="a" * 1100)
-    def testFilenameHandlingFallsBackWhenGeneratedPathExceedsDbLimit(self):
+    @mock.patch("documents.consumer.generate_unique_filename")
+    def testFilenameHandlingFallsBackWhenGeneratedPathExceedsDbLimit(self, m):
+        m.side_effect = lambda doc, archive_filename=False: Path(
+            ("a" * 1100 + ".pdf") if not archive_filename else ("b" * 1100 + ".pdf"),
+        )
+
         with self.get_consumer(
             self.get_test_file(),
             DocumentMetadataOverrides(title="new docs"),

--- a/src/documents/tests/test_consumer.py
+++ b/src/documents/tests/test_consumer.py
@@ -633,6 +633,29 @@ class TestConsumer(
 
         self._assert_first_last_send_progress()
 
+    @override_settings(FILENAME_FORMAT="a" * 1100)
+    def testFilenameHandlingFallsBackWhenGeneratedPathExceedsDbLimit(self):
+        with self.get_consumer(
+            self.get_test_file(),
+            DocumentMetadataOverrides(title="new docs"),
+        ) as consumer:
+            consumer.run()
+
+        document = Document.objects.first()
+        self.assertIsNotNone(document)
+        assert document is not None
+
+        self.assertEqual(document.filename, f"{document.pk:07d}.pdf")
+        self.assertLessEqual(len(document.filename), 1024)
+        self.assertLessEqual(
+            len(document.archive_filename),
+            1024,
+        )
+        self.assertIsFile(document.source_path)
+        self.assertIsFile(document.archive_path)
+
+        self._assert_first_last_send_progress()
+
     @override_settings(FILENAME_FORMAT="{correspondent}/{title}")
     @mock.patch("documents.signals.handlers.generate_unique_filename")
     def testFilenameHandlingUnstableFormat(self, m):

--- a/src/documents/tests/test_file_handling.py
+++ b/src/documents/tests/test_file_handling.py
@@ -1699,6 +1699,21 @@ class TestCustomFieldFilenameUpdates(
         self.assertTrue(Path(self.doc.source_path).is_file())
         self.assertLessEqual(m.call_count, 1)
 
+    @override_settings(FILENAME_FORMAT=None)
+    def test_overlong_storage_path_keeps_existing_filename(self):
+        initial_filename = generate_filename(self.doc)
+        Document.objects.filter(pk=self.doc.pk).update(filename=str(initial_filename))
+        self.doc.refresh_from_db()
+        Path(self.doc.source_path).parent.mkdir(parents=True, exist_ok=True)
+        Path(self.doc.source_path).touch()
+
+        self.doc.storage_path = StoragePath.objects.create(path="a" * 1100)
+        self.doc.save()
+
+        self.doc.refresh_from_db()
+        self.assertEqual(Path(self.doc.filename), initial_filename)
+        self.assertTrue(Path(self.doc.source_path).is_file())
+
 
 class TestPathDateLocalization:
     """

